File: app/src/main/java/com/vibhorsrv/cameraids/api/CameraIDs.java
package com.vibhorsrv.cameraids.api;

import android.hardware.camera2.CameraCharacteristics;

import com.vibhorsrv.cameraids.model.CameraModel;

import java.util.List;

/**
 * Created by Vibhor Srivastava
 */
public interface CameraIDs {
    interface Finder<T extends Iterable<? super CameraModel>> {
        void init();

        CameraCharacteristics getCameraCharacteristics(int cameraId);

        T getCameraModels();

        List<String> getApiCameraIdList();

        List<String> getAllCameraIdList();
    }

    interface Identifier<T extends Iterable<? super CameraModel>> {
        void init();

        void identifyCamera(T cameraModels);
    }

    interface Saver {
        void saveText(String path, String text);
    }
}

File: app/src/main/java/com/vibhorsrv/cameraids/api/CameraIDs.kt
package com.vibhorsrv.cameraids.api

import android.hardware.camera2.CameraCharacteristics
import com.vibhorsrv.cameraids.model.CameraModel

/**
 * Created by Vibhor Srivastava
 */
interface CameraIDs {
    interface Finder<T : Iterable<CameraModel>> {
        fun init()
        fun getCameraCharacteristics(cameraId: Int): CameraCharacteristics?
        val cameraModels: T
        val apiCameraIdList: List<String>
        val allCameraIdList: List<String>
    }

    interface Identifier<T : Iterable<CameraModel>> {
        fun init()
        fun identifyCamera(cameraModels: T)
    }

    interface Saver {
        fun saveText(path: String, text: String)
    }
}
